service/matching/taskWriter.go

```diff
@@ -218,8 +215,14 @@ func (w *taskWriter) appendTasks(
 	}
 }
 
-func (w *taskWriter) taskWriterLoop() {
-	w.initReadWriteState()
+func (w *taskWriter) taskWriterLoop(ctx context.Context) error {
+	err := w.initReadWriteState(ctx)
+	if err != nil {
+		if w.tlMgr.errShouldUnload(err) {
```

Comment:
This looks like a change in behavior from previous logic.  We used to always call `signalFatalProblem` when error during taskWriterLoop and now it seems to happen after checking err is the one which should result in Unload of tasklist.
Do we have a background pump trying to acquire TaskQueue lease?  What does it means to continue with the pump if initReadWriteState fails?

Reply:
Prior to this change we attempted to do the init in an infinite loop that only broke out if (1) we initialized or (2) we saw another shard owner. The problem was that such a loop would never break for things like calling taskWriter.Stop() so shutdown would hang. This is essentially the same except that now it's possible to interrupt init via context cancellation. In other words if taskWriter.Stop() is called we will no longer leak the goro and in such cases it will just exit normally.

Reply:
I'm not sure I get the `init in an infinite loop`.  Do you mean that we retry persistence errors infinitely?  The concern I have is that `initReadWriteState` could return an error which is not `errShouldUnload` resulting in taskWriter pump to be created.  As oppose to before we would always shutdown the pump on all errors.

Reply:
Yes, I'm referring to how we retry transient persistence errors. And how previously we did so _with no way to stop_. 

`initReadWriteState` returns an error in two conditions: (1) non-retryable persistence error (this is what happens if another process takes ownership of the queue-partition) or (2) context cancellation. In case #1 if the resilient matching feature flag is off then that's a fatal problem so we signal. In both cases, we return the error (line 224) rather than enter into the writerLoop. Entering the writerLoop with uninitialized state would probably be very bad and weird things would happen.

Reply:
Got it.  Makes sense.